src/isolate.rs

```diff
@@ -282,23 +289,77 @@ impl Isolate {
     }
   }
 
+  /// Runs the default MicrotaskQueue until it gets empty.
+  /// Any exceptions thrown by microtask callbacks are swallowed.
+  pub fn run_microtasks(&mut self) {
+    unsafe { v8__Isolate__RunMicrotasks(self) }
+  }
+
+  /// Enqueues the callback to the default MicrotaskQueue
+  pub fn enqueue_microtask(&mut self, microtask: Local<Function>) {
+    unsafe { v8__Isolate__EnqueueMicrotask(self, microtask) }
+  }
+
+  /// Disposes the isolate.  The isolate must not be entered by any
+  /// thread to be disposable.
+  unsafe fn dispose(&mut self) {
+    IsolateHandle::dispose(self);
+    v8__Isolate__Dispose(self)
+  }
+}
+
+// TODO(ry) Use AtomicPtr? Use Cell? Only one thread ever modifies it.
+/// Thread safe reference to an Isolate. This handle will out-live the Isolate
+/// itself, when that happens all methods on it will return false.
+pub struct IsolateHandle(Arc<Mutex<Option<*mut Isolate>>>);
+
+unsafe impl Send for IsolateHandle {}
+unsafe impl Sync for IsolateHandle {}
+
+impl IsolateHandle {
+  fn dispose(isolate: &mut Isolate) {
+    let slot_ptr = isolate.get_data(0) as *const Mutex<Option<*mut Isolate>>;
+    if !slot_ptr.is_null() {
+      unsafe { isolate.set_data(0, std::ptr::null_mut()) };
+      let handle = unsafe { Arc::from_raw(slot_ptr) };
+      let some_ptr = handle.lock().unwrap().take();
+      assert!(some_ptr.is_some());
+    }
+  }
+
+  fn new(isolate: &mut Isolate) -> Self {
+    let slot_ptr = isolate.get_data(0) as *mut Mutex<Option<*mut Isolate>>;
+    if slot_ptr.is_null() {
+      let x: Option<*mut Isolate> = Some(&mut *isolate);
```

Comment:
It would be good to see this through and implement Global::drop() using the reference count you get with this.

Reply:
What's the advantage of creating this thing lazily? It makes hard to reason about with virtually no benefits. 
Note that creating an Isolate is a quite expensive operation. Allocating an Arc certainly is not going to register on any benchmark.

Reply:
I want to make sure that in the simple case (no threads) there's zero overhead

Reply:
Versus 0.00000.1% overhead? 
Premature optimization much?

Reply:
Ideally I'll be able to strace deno and see no mutex locks being made. I think it's worth the one line it takes.